src/workerd/jsg/jsg.c++

```diff
@@ -41,10 +41,57 @@ const char* JsExceptionThrown::what() const noexcept {
   return whatBuffer.cStr();
 }
 
+Data::Data(v8::Isolate* isolate, v8::Local<v8::Data> handle)
+    : isolate(IsolateBase::from(isolate).getIsolatePtr()), handle(isolate, handle) {}
+
+Data::Data(Data&& other): isolate(other.isolate), handle(kj::mv(other.handle)) {
+  KJ_IF_MAYBE(t, other.tracedHandle) {
+    // `other` is a traced `Data`, but once moved, we don't assume the new location is traced.
+    // So, we need to make the handle strong.
+    handle.ClearWeak();
+
+    // Presumably, `other` is about to be destroyed. The destructor of `TracedReference`, though,
+    // does nothing, because it doesn't know if the reference is even still valid, since it
+    // could be called during GC sweep time. But here, we know that `other` is definitely still
+    // valid, because we wouldn't be moving from an unreachable object. So we should Reset() the
+    // `TracedReference` so that V8 knows it's gone, which might make minor GCs more effective.
+    t->Reset();
+
+    other.tracedHandle = nullptr;
+  }
+  other.isolate.clear();
+  assertInvariant();
+  other.assertInvariant();
+}
+
+Data::~Data() noexcept(false) {
+  destroy();
+}
+
+Data& Data::operator=(Data&& other) {
+  if (this != &other) {
+    destroy();
+    isolate = other.isolate;
+    handle = kj::mv(other.handle);
+    other.isolate.clear();
+    KJ_IF_MAYBE(t, other.tracedHandle) {
+      handle.ClearWeak();
+      t->Reset();
+      other.tracedHandle = nullptr;
+    }
+  }
+  assertInvariant();
+  other.assertInvariant();
+  return *this;
+}
+
 void Data::destroy() {
   assertInvariant();
-  if (isolate != nullptr) {
-    if (v8::Locker::IsLocked(isolate)) {
+  auto i = isolate.get();
+  // isolate.get() can return nullptr if Data::destroy() is called while or after the
+  // associated IsolateBase has been destroyed and the v8::Isolate has been disposed.
+  if (i != nullptr) {
```

Comment:
We can't just skip this logic. If we don't hold the isolate lock, then it is not safe to destroy `handle` in this thread. But when we return from `destroy()` to `~Data`, the handle is going to be destroyed.

Indeed, if the isolate has already been destroyed, then destroying `handle` is going to be UB and there's no way around it.

Unfortunately I think this approach just doesn't actually provide the protection that was hoped for. We need a different approach here.

Reply:
Hmmm yeah I was kind of afraid that would be the case :-/. I think for now I'll just close this and keep stewing over it. It would be ideal if we could have a workable safety rail here but currently not a massively pressing issue.